### PR TITLE
Java: Framework mode source candidates

### PR DESCRIPTION
**Pull request by kaeluka**
This adds support for source candidates, as well as +/- examples in framework mode.

The changes are similar to the recent PR (https://github.com/github/codeql/pull/14162) that did the same thing for application mode.

There's new kinds of endpoints:
 - a parameter of a method that may be overridden in a different code base
 - a qualifier of a method that may be overridden in a different code base
 - a return value of a method that may be called in a different code base

I've added some minor clean-up changes, by removing a module that did not do anything other than delegating to the `endpoint::getEnclosingCallable/0` implementation, and by removing the somewhat obscure (and unused) candidate-message concatenation.

## Reviewer

I tried to keep the individual commits as small as possible, this may be helpful in your review

### CODE REVIEW

**Comment by kaeluka** [comment]
@tausbn: We've talked about maybe removing that whole "erroneousEndpoints" business with this PR as a drive-by change, but haven't decided what to do. My opinion is that we should — what do you think?

**Comment by kaeluka** [comment]
@tausbn, as we discussed recently, I've pushed a commit that removes the predicate. I'll now write an issue to upgrade to inline expectation tests.

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll [hunk 1]
```diff
@@ -272,76 +383,5 @@ private class NonPublicMethodCharacteristic extends CharacteristicsImpl::Uninter
...
- * have some overlap in their results. This indicates a problem with the underlying Java modeling. Specifically,
- * `PathCreation` is prone to FPs.
- */
-private predicate knownOverlappingCharacteristics(
-  EndpointCharacteristic characteristic1, EndpointCharacteristic characteristic2
-) {
-  characteristic1 != characteristic2 and
-  characteristic1 = ["mad taint step", "create path", "read file", "known non-sink"] and
-  characteristic2 = ["mad taint step", "create path", "read file", "known non-sink"]
+  override predicate appliesToEndpoint(Endpoint e) { not e.getEnclosingCallable().isPublic() }
```
Comment (tausbn) [inline java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:386]: I'm slightly confused about the interplay between this an the use of `ModelExclusions::ModelApi` elsewhere (e.g. the IPA type for sources, and `NotAModelApi`). I feel like we're checking `isPublic` in more than one place (though I guess not everything is assumed to be an instance of `ModelApi`).

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll [hunk 2]
```diff
@@ -25,7 +25,23 @@ newtype JavaRelatedLocationType =
 
 newtype TFrameworkModeEndpoint =
   TExplicitParameter(Parameter p) or
-  TQualifier(Callable c)
+  TQualifier(Callable c) or
+  TReturnValue(Callable c) or
+  TOverridableParameter(Method m, Parameter p) {
+    p.getCallable() = m and
+    m instanceof ModelExclusions::ModelApi and
+    m.fromSource() and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:33]: 👍 thx!

[the remark this answers is not in the document]
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll]: [6be64a6](https://github.com/github/codeql/pull/14197/commits/6be64a6dbdf9831c2a4c11149bb70cf662ab5b5a)

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll [hunk 3]
```diff
@@ -25,7 +25,23 @@ newtype JavaRelatedLocationType =
...
+    not m.getDeclaringType().isFinal() and
+    not m.isFinal() and
+    not m.isStatic()
+  } or
+  TOverridableQualifier(Method m) {
+    m instanceof ModelExclusions::ModelApi and
+    m.fromSource() and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll]: [6be64a6](https://github.com/github/codeql/pull/14197/commits/6be64a6dbdf9831c2a4c11149bb70cf662ab5b5a)

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll [hunk 4]
```diff
@@ -76,11 +114,81 @@ class QualifierEndpoint extends FrameworkModeEndpoint, TQualifier {
...
+  override string getExtensibleType() { result = "sinkModel" }
+}
+
+class ReturnValue extends FrameworkModeEndpoint, TReturnValue {
+  Callable callable;
+
+  ReturnValue() { this = TReturnValue(callable) and callable.fromSource() }
+
+  override int getIndex() {
+    // FIXME bogus value
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll]: you're right!! It felt GREAT to delete that: 
[427d3fb](https://github.com/github/codeql/pull/14197/commits/427d3fbe14c892ef246772b1f98950a78fb74901)

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll [hunk 5]
```diff
@@ -25,7 +25,23 @@ newtype JavaRelatedLocationType =
 
 newtype TFrameworkModeEndpoint =
   TExplicitParameter(Parameter p) or
-  TQualifier(Callable c)
+  TQualifier(Callable c) or
+  TReturnValue(Callable c) or
+  TOverridableParameter(Method m, Parameter p) {
+    p.getCallable() = m and
+    m instanceof ModelExclusions::ModelApi and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:32]: If you look at the `ModelExclusions` module, you'll see that almost all code in it is about _excluded_ models. It just so happens that the one thing we use from it is the single piece of code that includes a negation :)

Renaming the module: that'd be a discussion with the Java team that I'd rather not have in the PR. We could rename the IMPORT like `private import semmle.code.java.dataflow.internal.ModelExclusions as ModelInclusions` but that's just.. wrong 😄

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll (hunk 1 above)
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: Dropping this, iow, not adding this characteristic on top would reflect the view that the language team (maintainers of `ModelApi`) are the real experts for what is worth having a model for.

Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: 👀 @jhelie

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll (hunk 5 above)
[the remark this answers is not in the document]
Comment (tausbn) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:32]: Ah, fair enough. I somehow thought this was one of _our_ files. In that case, I think I can probably live with it. (I also think `ModelApi` is a name that says basically nothing about what it actually is, but whatever. Naming things is hard.)

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll (hunk 1 above)
Comment (tausbn) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: I would be in favour of dropping it. 👍

Comment (tausbn) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: I would be in favour of dropping it. 👍

**Review by tausbn — APPROVED** [review]
Really nice work!

File: java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll (hunk 1 above)
Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: It doesn't sound like something that should make up for a huge increase. Some data points:

`apache/geode`: 103625 candidates ➡️ 114467 candidates (+10%)
`square/okhttp`:  11841 candidates ➡️ 11867 candidates (+0,2%)

Sounds good, @jhelie?

Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: Btw, getting more candidates is a _good_ thing, it gives us more complete models 🙂 👍

Comment (jhelie) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: Yes I think we can live with a 10% increase 👍 

> Btw, getting more candidates is a good thing, it gives us more complete models 🙂 👍

Agreed as long as it's not +10000%, we're hammering the API enough as it is 😅

Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll:366]: haha, yes. that would raise some concern, agreed 👍

Comment (kaeluka) [reply java/ql/automodel/src/AutomodelFrameworkModeCharacteristics.qll]: [fdbc553](https://github.com/github/codeql/pull/14197/commits/fdbc55302933c7c56f42ce19cc038dc294744496) ✅

**Review by tausbn — APPROVED** [review]

**Pull request merged**